sw/host/opentitanlib/src/transport/ti50emulator/i2c.rs

```diff
@@ -2,14 +2,268 @@
+    }
+
+    /// Receive write response
+    pub fn recv_write_response(fd: &mut UnixStream, deadline: Instant) -> Result<()> {
+        let buf = &mut [1u8; 1];
+        let _rx_len = Ti50I2cBus::rx(fd, &mut buf[..], deadline)?;
+        match buf {
+            [TI50_I2C_BUS_WRITE_RES] => Ok(()),
+            [raw] => Err(
+                I2cError::Generic(format!("Unexpected BUS control message: {:02X}", raw)).into(),
+            ),
+        }
+    }
+
+    /// Receive read response
+    pub fn recv_read_response(fd: &mut UnixStream, deadline: Instant) -> Result<()> {
+        let buf = &mut [1u8; 1];
+        let _rx_len = Ti50I2cBus::rx(fd, &mut buf[..], deadline)?;
+        match buf {
+            [TI50_I2C_BUS_READ_RES] => Ok(()),
+            [raw] => Err(
+                I2cError::Generic(format!("Unexpected BUS control message: {:02X}", raw)).into(),
+            ),
+        }
+    }
+}
+
+/// Structure representing Emulated I2C BUS
+pub struct Ti50I2cBus {
+    inner: Rc<RefCell<Inner>>,
+    // This socket is valid as long as SubProcess is running.
+    socket: RefCell<Option<UnixStream>>,
+    // full path to socket file
+    path: PathBuf,
+    // last SubProcess ID
+    last_id: Cell<u64>,
+}
+
+impl Ti50I2cBus {
+    /// Create new instance of [`Ti50I2cBus`] based on provided parameters.
+    pub fn open(ti50: &Ti50Emulator, path: &str) -> Result<Self> {
+        let soc_path = ti50.inner.borrow().process.get_runtime_dir().join(path);
+        Ok(Self {
+            inner: Rc::clone(&ti50.inner),
+            socket: RefCell::default(),
+            path: soc_path,
+            last_id: Cell::new(EMULATOR_INVALID_ID),
+        })
+    }
+
+    /// Function re-connect socket to `SubProcess` when detect
+    /// that process was restarted.
+    pub fn reconnect(&self) -> Result<()> {
+        let mut socket = self.socket.borrow_mut();
+        let id = self.inner.borrow_mut().process.get_id();
+        if self.last_id.get() != id {
+            *socket = Some(UnixStream::connect(&self.path)?);
+            self.last_id.set(id);
+        }
+        Ok(())
+    }
+
+    /// Function check if I2C transaction should be performed on DUT.
+    pub fn check_state(&self) -> Result<()> {
+        let process = &mut self.inner.borrow_mut().process;
+        process.update_status()?;
+        match process.get_state() {
+            EmuState::On => Ok(()),
+            state => Err(I2cError::Generic(format!(
+                "Operation not supported in Emulator state: {}",
+                state
+            ))
+            .into()),
+        }
+    }
+
+    /// Function try to receive data from unix socket.
+    pub fn rx(fd: &mut UnixStream, data: &mut [u8], deadline: Instant) -> Result<usize> {
+        let mut rx_count: usize = 0;
+        while rx_count < data.len() {
+            if Instant::now() > deadline {
+                bail!(I2cError::Timeout);
+            }
+            match fd.read(&mut data[rx_count..]) {
```

Comment:
I think this will work, but be wasteful of CPU cycles, in that until data arrives, it will invoke the `read()` syscall as fast as the CPU can do, each time returning immediately with `ErrorKind::WouldBlock`.  I am not sure how to do it in Rust, but in traditional C, you would replace the deadline comparison above with a call to `select()`, providing a timeout parameter equal `deadline - Instant::now()`.  (The same goes for the `tx()` function.)

Reply:
Standard library provide set_read/write_timeout but it is set per operation. 
Also we have some read/write with timeout already implemented in src/util/file.rs but....
There is a difference between the time of the write/read itself and the time of the entire transaction.
Current deadline measure total time spend in transaction with is better.

About CPU usage, if test fail by timeout you already waste a lot of time in CI and this busy loop will be only small fraction of it.  ;)

I can rewrite the code to use pool / select if you know the timeout should be on a scale of seconds.

Reply:
I see.  We have several options, then:
1) Keep the deadline and loop, but set a "short" timeout of 100ms or so on each read/write operation in the loop.  That way, we can overrun the deadline by at most a tenth of a second, and we do not waste nearly as many CPU cycles as with the current unthrottled loop.
2) Invoke set_(read|write)_timeout on each iteration of the loop, based on how much time is left until the deadline.
3) Add poll()/select() into the loop.

I do want you to implement one of the three above, as I am afraid that the current code could eat so many CPU cycles that it slows down the subprocess in the other end, that we are really waiting for.

Reply:
I will try option to implement (2).